### PR TITLE
feat(usage): add support resource arrays in usage-file

### PR DESCRIPTION
**Pull request by ekifel**
Issue #528 

Example: 
Usage-file: 
```
 aws_cloudwatch_log_group.my_log_group[*]:
    storage_gb: 1000               
    monthly_data_ingested_gb: 1000 
    monthly_data_scanned_gb: 200  
``` 

Terraform config: 
```
resource "aws_cloudwatch_log_group" "my_log_group" {
  count = 2
  name              = "log-group${count.index}"
}
```

Output: 
```
  NAME                                          MONTHLY QTY  UNIT  PRICE   HOURLY COST  MONTHLY COST  

  aws_cloudwatch_log_group.my_log_group[0]                                                            
  ├─ Data ingested                                    1,000  GB    0.5000       0.6849      500.0000  
  ├─ Archival Storage                                 1,000  GB    0.0300       0.0411       30.0000  
  └─ Insights queries data scanned                      200  GB    0.0050       0.0014        1.0000  
  Total                                                                         0.7274      531.0000  
                                                                                                      
  aws_cloudwatch_log_group.my_log_group[1]                                                            
  ├─ Data ingested                                    1,000  GB    0.5000       0.6849      500.0000  
  ├─ Archival Storage                                 1,000  GB    0.0300       0.0411       30.0000  
  └─ Insights queries data scanned                      200  GB    0.0050       0.0014        1.0000  
  Total                                                                         0.7274      531.0000
```

### CODE REVIEW

File: internal/providers/terraform/parser.go [hunk 1]
```diff
@@ -99,7 +99,12 @@ func (p *Parser) parseJSONResources(parsePrior bool, baseResources []*schema.Res
 	p.stripDataResources(resData)
 
 	for _, d := range resData {
-		if r := p.createResource(d, usage[d.Address]); r != nil {
+		usageAddress := usage[d.Address]
+
+		if arrayUsageAddress := usage[fmt.Sprintf("%s[*]", strings.Split(d.Address, "[")[0])]; arrayUsageAddress != nil {
```
Comment (aliscott) [inline internal/providers/terraform/parser.go:104]: Would this work with counts in modules? If the module itself uses `count` or `for_each` and the resource has the same then I think the address comes through as `module.my_module[0].aws_instance.my_instance[2]`

**Review by aliscott — COMMENTED** [review]
Awesome work! The tests look great.

File: infracost-usage-example.yml [hunk 2]
```diff
@@ -14,6 +14,16 @@ resource_usage:
   # module.lambda_function.aws_lambda_function.this[0]:
   #   monthly_requests: 20000
   #   request_duration_ms: 600
+  #
+  # You can use [*] for array of resources to apply usage-values individually to each element of the array.
+  # Specifying an array element have priority.
```
Comment (alikhajeh1) [inline infracost-usage-example.yml:19]: If this what we mean here?
```
If usage for an array element (e.g. this[0]) and [*] exist, only the array element's usage will be applied to that resource.
```

Comment (ekifel) [reply infracost-usage-example.yml:19]: Yes it is. Should I fix it to your describe?

Comment (alikhajeh1) [reply infracost-usage-example.yml:19]: Thanks, that would be good.

**Review by aliscott — APPROVED** [review]

File: internal/providers/terraform/parser_test.go [hunk 3]
```diff
@@ -0,0 +1,300 @@
+package terraform
```
Comment (aliscott) [inline internal/providers/terraform/parser_test.go:1]: We already have `parser_internal_test.go`, I think this is left from when we had integration tests for the parser. Do you think we should move all the tests from there into this `parser_test.go` file as well?

File: internal/providers/terraform/parser.go [hunk 4]
```diff
@@ -100,7 +100,16 @@ func (p *Parser) parseJSONResources(parsePrior bool, baseResources []*schema.Res
 	p.stripDataResources(resData)
 
 	for _, d := range resData {
-		if r := p.createResource(d, usage[d.Address]); r != nil {
+		var usageData *schema.UsageData
+
+		if ud := usage[d.Address]; ud != nil {
+			usageData = ud
+		} else if strings.HasSuffix(d.Address, "]") {
+			if arrayUsageData := usage[fmt.Sprintf("%s*]", d.Address[:len(d.Address)-2])]; arrayUsageData != nil {
```
Comment (aliscott) [inline internal/providers/terraform/parser.go:108]: Does this work when count >= 10?

**Review by aliscott — COMMENTED** [review]
@ekifel, looking good. Just a couple more questions on this.

File: internal/providers/terraform/parser.go (hunk 4 above)
Comment (ekifel) [reply internal/providers/terraform/parser.go:108]: Do you mean 'count' as lenght of d.Address? If yes, sure. I checked it in situations with arrays of modules, it was over ~20-30 symbols and it works fine.

File: internal/providers/terraform/parser_test.go (hunk 3 above)
Comment (ekifel) [reply internal/providers/terraform/parser_test.go:1]: I think we should, I don't see any reasons to split that tests to two files. Should I move them?

Comment (aliscott) [reply internal/providers/terraform/parser_test.go:1]: Yes that would be great, thanks.

File: internal/providers/terraform/parser.go (hunk 4 above)
Comment (aliscott) [reply internal/providers/terraform/parser.go:108]: I mean where `count = 11` in the resource. So the resource would have `aws_instance.my_instance[10]` as the address.

Comment (ekifel) [reply internal/providers/terraform/parser.go:108]: ohh, missunderstanding, forgot about that. I will add checking an array lenght before it.

Comment (ekifel) [reply internal/providers/terraform/parser.go]: Fixed! Ready to review

**Comment by alikhajeh1** [comment]
@ekifel I tested this, love it!

I found one array case where it's not working though, can you please take a look?
Given this terraform resource:
```
resource "google_container_cluster" "with_node_pools_zonal" {
  name     = "with-node-pools"
  location = "us-central1-a"

  node_pool {
    node_config {
      machine_type = "n1-standard-16"
    }
  }

  node_pool {
    node_config {
      machine_type  = "n1-standard-16"
    }
  }
}
```

And this usage-file:
```
version: 0.1
resource_usage:
  google_container_cluster.with_node_pools_zonal:
    node_pool[*]:
      nodes: 8
```

If I do `node_pool[0]` it works as expected. It doesn't apply the usage-file wildcard.

**Pull request merged**